### PR TITLE
13027 - cache the SSLContext and SSLSocketFactory

### PR DESCRIPTION
**Pull request by WhiteCat22**
Cache the SSLContext and the SSLSocket Factory, then get the SSLContext from JSSEHelper and compare against our cached copy each request to determine if we need to refresh our cached SSLSocketFactory

We need to do this because JSSEHelper only caches the SSLContext. Calling SSLContext.getSocketFactory() creates a new SSLSocketFactory every time. So, we need to cache the socket factory for re-use, but also the SSLContext so that we can compare it against the current SSLContext. If they are not equal, then the SSL config has changed and we need to get the new SSLContext and cache the new SSLSocketFactory to support dynamic SSL updates.

#14934

### CODE REVIEW

**Comment by WhiteCat22** [comment]
#run-libby-bot
#build

**Review by acdemyers — APPROVED** [review]

File: dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java [hunk 1]
```diff
@@ -28,6 +29,9 @@
 public class JaxRsSSLManager {
     private static final TraceComponent tc = Tr.register(JaxRsSSLManager.class);
 
+    private static final Map<String, SSLSocketFactory> socketFactories = new HashMap<>();
+    private static SSLContext sslContext = null;
```
Comment (andymc12) [inline dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java:33]: Wouldn't we want this to be a Map too? (Map<String, SSLContext>) It seems like we might have more than one SSLContext configured for a given server, and if we just have one static field, then we might always skip the cache (i.e. app1 uses context1, then app2 uses context2 forcing a swap of the sslContext field, etc.).

File: dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java [hunk 2]
```diff
@@ -78,8 +82,23 @@ public Boolean run() throws SSLException {
...
+            SSLContext context = jsseHelper.getSSLContext(connectionInfo, sslProps);
+
+            boolean recache = false;
+            if (sslContext == null || !context.equals(sslContext)) {
+                // SSL config has changed, re-cache the SSLSocketFactory
+                synchronized (sslContext) {
+                    sslContext = context;
+                }
+                recache = true;
+            }
```
Comment (andymc12) [inline dev/com.ibm.ws.jaxrs.2.0.security/src/com/ibm/ws/jaxrs20/appsecurity/security/JaxRsSSLManager.java:94]: see above comments about using a Map for sslContext.

**Review by andymc12 — CHANGES_REQUESTED** [review]

**Comment by WhiteCat22** [comment]
#run-libby-bot
#build

**Review by andymc12 — APPROVED** [review]
LGTM - thanks @WhiteCat22!

**Comment by WhiteCat22** [comment]
Testing is complete, this solved the issue of persistent TLS connections breaking while also supporting dynamic ssl config changes without having to restart the server.

**Pull request merged**